File: assets/js/monitor/momentum.js
(function () {

  const Monitor = window.Monitor;

  const { app, el } = Monitor;



  Monitor.signalLabels = {

    strong_buy: "强多",

    buy: "做多",

    neutral: "观望",

    sell: "做空",

    strong_sell: "强空",

  };



  Monitor.momentum = {

    hu: { last: "neutral", strength: 0 },

    co: { last: "neutral", strength: 0 },

    au: { last: "neutral", strength: 0 },

    cg: { last: "neutral", strength: 0 },

    btc: { last: "neutral", strength: 0 },

  };

  // Cooldown tracking per symbol: { symbol: remainingBars }
  Monitor._cooldowns = {};

  /**
   * 按时间窗口入队：同一窗口内覆写末条价格（取最新值），跨窗口时追加新 bar。
   * windowMs 读取 Monitor.constants.BAR_WINDOW_MS（由 monitor.config.json frontend.bar_window_ms 设置），缺省 30000ms。
   */
  Monitor._pushByTimeWindow = function (arr, price, ts, cap) {
    const windowMs = (Monitor.constants && Monitor.constants.BAR_WINDOW_MS) || 30000;
    if (arr.length > 0 && ts - arr[arr.length - 1].t < windowMs) {
      arr[arr.length - 1].y = price;
    } else {
      arr.push({ t: ts, y: price });
      while (arr.length > cap) arr.shift();
    }
  };

  /** RSI 超买/超卖修正信号（与后端 _fuse_with_rsi 对齐） */
  function _fuseWithRSI(sig, rsi) {
    if (sig === "buy" && rsi > 70) return "neutral";
    if (sig === "sell" && rsi < 30) return "neutral";
    return sig;
  }

  /** BB 融合修正 EMA 动量信号（与后端 _fuse_with_bb 对齐） */

  function _fuseWithBB(baseSig, pctB, bwExpanding, buyKill, sellKill) {

    const bk = buyKill != null ? buyKill : 0.3;

    const sk = sellKill != null ? sellKill : 0.7;

    let sig = baseSig;

    if (sig === "buy") {

      if (pctB < bk) sig = "neutral";

      else if (pctB > 0.5 && bwExpanding) sig = "strong_buy";

    } else if (sig === "strong_buy") {

      if (pctB > 1.0) sig = "buy";

    } else if (sig === "sell") {

      if (pctB > sk) sig = "neutral";

      else if (pctB < 0.5 && bwExpanding) sig = "strong_sell";

    } else if (sig === "strong_sell") {

      if (pctB < 0.0) sig = "sell";

    }

    return sig;

  }



  /** EMA 短/长 张口 + 短 EMA 斜率 + Bollinger 带融合 */

  Monitor.calcMomentum = function (series, shortP, longP, thresholds) {

    if (!series || series.length < longP * 2) return null;

    const vals = series.map(p => p.y);

    const emaS = Monitor.ema(vals, shortP);

    const emaL = Monitor.ema(vals, longP);

    const lastS = emaS[emaS.length - 1];

    const lastL = emaL[emaL.length - 1];

    const prevS = emaS[emaS.length - 2];

    if (lastS == null || lastL == null || prevS == null) return null;

    const spreadPct = lastL !== 0 ? ((lastS - lastL) / lastL) * 100 : 0;

    const slopePct = prevS !== 0 ? ((lastS - prevS) / prevS) * 100 : 0;



    const th = thresholds || Monitor.getMomentumThresholds();

    const se = th.spreadEntry;

    const ss = th.spreadStrong;

    const sl = th.slopeEntry != null ? th.slopeEntry : 0.02;

    const sm = th.strengthMul != null ? th.strengthMul : 120;

    const bbP = th.bbPeriod || 0;

    const bbM = th.bbMult || 2.0;

    const bbBuyKill = th.bbBuyKill != null ? th.bbBuyKill : 0.3;

    const bbSellKill = th.bbSellKill != null ? th.bbSellKill : 0.7;



    let signal = "neutral";

    if (lastS > lastL && spreadPct > se && slopePct > sl) {

      signal = spreadPct > ss ? "strong_buy" : "buy";

    } else if (lastS < lastL && spreadPct < -se && slopePct < -sl) {

      signal = spreadPct < -ss ? "strong_sell" : "sell";

    }



    // Bollinger 带融合

    let bb = null;

    if (bbP > 0 && vals.length >= bbP) {

      const bbNow = Monitor.bollingerAt(vals, bbP, bbM);

      const bbPrev = vals.length > bbP ? Monitor.bollingerAt(vals.slice(0, -1), bbP, bbM) : null;

      if (bbNow) {

        const bwExpanding = bbPrev != null && bbNow.bandwidth > bbPrev.bandwidth;

        signal = _fuseWithBB(signal, bbNow.percentB, bwExpanding, bbBuyKill, bbSellKill);

        // Squeeze detection: bandwidth <= bb_period-bar min
        let squeeze = false;
        if (vals.length >= bbP * 2) {
          let minBW = Infinity;
          for (let j = Math.max(0, vals.length - bbP - 1); j < vals.length - 1; j++) {
            const bj = Monitor.bollingerAt(vals.slice(0, j + 1), bbP, bbM);
            if (bj) minBW = Math.min(minBW, bj.bandwidth);
          }
          if (bbNow.bandwidth <= minBW) squeeze = true;
        }

        bb = { ...bbNow, bwExpanding, squeeze };

      }

    }



    // RSI fusion
    const rsiP = th.rsiPeriod || 0;
    let rsi = null;
    if (rsiP > 0 && vals.length >= rsiP + 1) {
      rsi = Monitor.rsiAt(vals, rsiP);
      if (rsi != null) signal = _fuseWithRSI(signal, rsi);
    }

    return {

      signal,

      spreadPct,

      slopePct,

      shortEMA: lastS,

      longEMA: lastL,

      strength: Math.min(100, Math.abs(spreadPct) * sm),

      bb,

      rsi,

    };

  };



  Monitor.renderSignal = function (prefix, info, decimals) {

    const symbolByPrefix = {
      hu: "huyin",
      co: "comex",
      au: "hujin",
      cg: "comex_gold",
      btc: "btc",
    };

    // 更新 EMA 周期标签（始终更新，即使 info 为 null）
    const symbol = symbolByPrefix[prefix];
    if (symbol && Monitor.getMomentumPeriods) {
      const periods = Monitor.getMomentumPeriods(symbol);
      if (periods) {
        const badgeNode = document.getElementById(prefix + "SignalBadge");
        let tagEl = null;
        if (badgeNode) {
          const card = badgeNode.closest(".signal-card");
          if (card) tagEl = card.querySelector(".ema-tag");
        }
        if (!tagEl) {
          const label = document.getElementById(prefix + "SigLabel");
          if (label) tagEl = label.querySelector(".ema-tag");
        }
        if (tagEl) {
          const current = tagEl.textContent;
          let suffix = "";
          if (current.includes("（")) suffix = current.slice(current.indexOf("（"));
          else if (current.includes("+Boll")) suffix = current.slice(current.indexOf("+Boll"));
          tagEl.textContent = `EMA${periods.shortP}/${periods.longP}${suffix}`;
        }
      }
    }

    const findEl = (...ids) => ids.map(id => el(id)).find(Boolean);

    const badge = findEl(prefix + "SignalBadge", prefix + "SigBadge");

    const slopeEl = findEl(prefix + "ROC", prefix + "SigSlope");

    const emaFastEl = findEl(prefix + "EmaF", prefix + "SigEmaS");

    const emaSlowEl = findEl(prefix + "EmaS", prefix + "SigEmaL");

    const pointsEl = findEl(prefix + "SigPts");

    const noteEl = findEl(prefix + "SigNote");

    const bar = findEl(prefix + "SigBar");



    if (!badge || !slopeEl || !emaFastEl || !emaSlowEl || !bar) return;



    if (!info) {

      badge.className = "signal-badge neutral";

      badge.textContent = "等待";

      slopeEl.textContent = "--";

      emaFastEl.textContent = "--";

      emaSlowEl.textContent = "--";

      bar.style.width = "50%";

      bar.className = "signal-bar-inner flat";

      if (pointsEl) pointsEl.textContent = "--";

      if (noteEl) {

        const symbol = symbolByPrefix[prefix];

        const lp = Monitor.getMomentumPeriods(symbol).longP ?? 20;

        const minPts = lp * 2;

        noteEl.textContent = `需至少 ${minPts} 个有效价格点后计算`;

      }

      return;

    }



    badge.className = "signal-badge " + info.signal.replace("_", "-");

    badge.textContent = Monitor.signalLabels[info.signal];

    const sl = info.slopePct;

    slopeEl.innerHTML = `<span class="val ${sl >= 0 ? "up" : "down"}">${sl >= 0 ? "+" : ""}${sl.toFixed(3)}%</span>`;

    emaFastEl.textContent = info.shortEMA.toFixed(decimals);

    emaSlowEl.textContent = info.longEMA.toFixed(decimals);

    bar.style.width = info.strength.toFixed(0) + "%";

    bar.className = "signal-bar-inner " + (info.signal.includes("buy") ? "bull" : info.signal.includes("sell") ? "bear" : "flat");

    if (pointsEl) pointsEl.textContent = "实时";

    // BB 指标渲染

    const bollBEl = findEl(prefix + "BollB");

    const bollBWEl = findEl(prefix + "BollBW");

    if (bollBEl && info.bb) {

      const b = info.bb.percentB;

      const cls = b > 0.8 ? "up" : b < 0.2 ? "down" : "";

      bollBEl.innerHTML = `<span class="val ${cls}">${b.toFixed(2)}</span>`;

    } else if (bollBEl) {

      bollBEl.textContent = "--";

    }

    if (bollBWEl && info.bb) {

      bollBWEl.textContent = info.bb.bandwidth.toFixed(2) + "%";

    } else if (bollBWEl) {

      bollBWEl.textContent = "--";

    }



    if (noteEl) {

      let note = `动量差 ${info.spreadPct >= 0 ? "+" : ""}${info.spreadPct.toFixed(3)}%`;

      if (info.bb) {

        const b = info.bb.percentB;

        if (info.bb.squeeze) note += " | Boll缩口";

        else if (b > 1.0) note += " | 超买";

        else if (b < 0.0) note += " | 超卖";

        else if (info.bb.bwExpanding) note += " | 带宽扩张";

      }

      noteEl.textContent = note;

    }

    // RSI 指标渲染
    const rsiEl = findEl(prefix + "RSI");
    if (rsiEl && info.rsi != null) {
      const r = info.rsi;
      const rcls = r > 70 ? "up" : r < 30 ? "down" : "";
      rsiEl.innerHTML = `<span class="val ${rcls}">${r.toFixed(1)}</span>`;
    } else if (rsiEl) {
      rsiEl.textContent = "--";
    }

  };



  Monitor.refreshMomentumLabels = function () {

    // 标签已由 renderSignal 根据品种特定周期动态更新，此处无需全局统一刷新。
    // 保留该函数以防止 core.js 调用时报错。

  };



  Monitor.playSignalTone = function (signal) {

    try {

      const enabled = localStorage.getItem("signalSound") !== "off";

      if (!enabled || !["strong_buy", "strong_sell"].includes(signal)) return;



      const Ctx = window.AudioContext || window.webkitAudioContext;

      if (!Ctx) return;

      app.audioCtx = app.audioCtx || new Ctx();

      if (app.audioCtx.state === "suspended") app.audioCtx.resume();



      const ctx = app.audioCtx;

      const osc = ctx.createOscillator();

      const gain = ctx.createGain();

      osc.connect(gain);

      gain.connect(ctx.destination);

      osc.type = signal === "strong_buy" ? "triangle" : "sawtooth";

      osc.frequency.value = signal === "strong_buy" ? 880 : 440;

      gain.gain.setValueAtTime(0.0001, ctx.currentTime);

      gain.gain.exponentialRampToValueAtTime(0.04, ctx.currentTime + 0.01);

      gain.gain.exponentialRampToValueAtTime(0.0001, ctx.currentTime + 0.18);

      osc.start(ctx.currentTime);

      osc.stop(ctx.currentTime + 0.2);

    } catch (_) {}

  };



  Monitor.updateMomentumSignals = function (data) {

    const backendSignals = data.signals || data._signals || Monitor.instrumentSignals || {};

    const hu = data.huyin;

    const co = data.comex;

    const huSeries = app.silverLivePoints.map(p => ({ x: p.t, y: p.y }));

    const coSeries = app.comexSilverLivePoints.map(p => ({ x: p.t, y: p.y }));

    // 使用品种特定的周期和阈值参数
    const huP = Monitor.getMomentumPeriods("huyin");

    const coP = Monitor.getMomentumPeriods("comex");

    const _valid = s => s && s.slopePct != null && s.shortEMA != null && s.longEMA != null;
    const huSig = _valid(backendSignals.ag0) ? backendSignals.ag0 : Monitor.calcMomentum(huSeries, huP.shortP, huP.longP, Monitor.getMomentumThresholds("huyin"));

    const coSig = _valid(backendSignals.xag) ? backendSignals.xag : Monitor.calcMomentum(coSeries, coP.shortP, coP.longP, Monitor.getMomentumThresholds("comex"));



    if (huSig && Monitor.momentum.hu.last !== huSig.signal) {

      Monitor.playSignalTone(huSig.signal);

      Monitor.momentum.hu = { last: huSig.signal, strength: huSig.strength };

    }

    if (coSig && Monitor.momentum.co.last !== coSig.signal) {

      Monitor.playSignalTone(coSig.signal);

      Monitor.momentum.co = { last: coSig.signal, strength: coSig.strength };

    }



    Monitor.renderSignal("hu", huSig, 1);

    Monitor.renderSignal("co", coSig, 3);

  };



  Monitor.updateGoldMomentumSignals = function (data) {

    const backendSignals = data.signals || data._signals || Monitor.instrumentSignals || {};

    const au = data.hujin;

    const cg = data.comexGold;

    const auSeries = (app.goldLivePoints || []).map(p => ({ x: p.t, y: p.y }));

    const cgSeries = (app.comexGoldLivePoints || []).map(p => ({ x: p.t, y: p.y }));

    // 使用品种特定的周期和阈值参数
    const auP = Monitor.getMomentumPeriods("hujin");

    const cgP = Monitor.getMomentumPeriods("comex_gold");

    const _valid2 = s => s && s.slopePct != null && s.shortEMA != null && s.longEMA != null;
    const auSig = _valid2(backendSignals.au0) ? backendSignals.au0 : Monitor.calcMomentum(auSeries, auP.shortP, auP.longP, Monitor.getMomentumThresholds("hujin"));

    const cgSig = _valid2(backendSignals.xau) ? backendSignals.xau : Monitor.calcMomentum(cgSeries, cgP.shortP, cgP.longP, Monitor.getMomentumThresholds("comex_gold"));



    if (auSig && Monitor.momentum.au.last !== auSig.signal) {

      Monitor.playSignalTone(auSig.signal);

      Monitor.momentum.au = { last: auSig.signal, strength: auSig.strength };

    }

    if (cgSig && Monitor.momentum.cg.last !== cgSig.signal) {

      Monitor.playSignalTone(cgSig.signal);

      Monitor.momentum.cg = { last: cgSig.signal, strength: cgSig.strength };

    }



    Monitor.renderSignal("au", auSig, 2);

    Monitor.renderSignal("cg", cgSig, 2);

  };

  Monitor.updateCryptoMomentumSignals = function (data) {
    const backendSignals = data.signals || data._signals || Monitor.instrumentSignals || {};
    const btc = data.btc;
    const btcSeries = (app.btcLivePoints || []).map(p => ({ x: p.t, y: p.y }));
    const btcP = Monitor.getMomentumPeriods("btc");
    const _valid3 = s => s && s.slopePct != null && s.shortEMA != null && s.longEMA != null;
    const btcSig = _valid3(backendSignals.btc) ? backendSignals.btc : Monitor.calcMomentum(btcSeries, btcP.shortP, btcP.longP, Monitor.getMomentumThresholds("btc"));

    if (btcSig && Monitor.momentum.btc.last !== btcSig.signal) {
      Monitor.playSignalTone(btcSig.signal);
      Monitor.momentum.btc = { last: btcSig.signal, strength: btcSig.strength };
    }

    Monitor.renderSignal("btc", btcSig, 2);
  };


  Monitor.refreshMomentumLabels();

})();

File: backend/strategies/momentum.py
"""动量信号（EMA 短/长张口 + 短 EMA 一步斜率 + Bollinger 带融合），与前端 momentum.js 对齐。"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Any


@dataclass
class MomentumParams:
    short_p: int = 5
    long_p: int = 20
    spread_entry: float = 0.10
    spread_strong: float = 0.35
    slope_entry: float = 0.02
    strength_multiplier: float = 120.0
    cooldown_bars: int = 0
    bb_period: int = 20
    bb_mult: float = 2.0
    rsi_period: int = 14  # 0 = disabled
    bb_buy_kill: float = 0.3   # buy 信号在 %B 低于此值时被压制为 neutral
    bb_sell_kill: float = 0.7  # sell 信号在 %B 高于此值时被压制为 neutral
    min_volatility_pct: float = 0.0  # 最小价格波动率(%)。>0 时若近期 CV 低于此值，方向信号降级为 neutral


def ema_series(values: list[float], period: int) -> list[float]:
    """EMA with SMA seed: uses average of first `period` values as initial seed to reduce cold-start bias."""
    if not values:
        return []
    k = 2.0 / (period + 1)
    n = len(values)
    if n >= period:
        seed = sum(values[:period]) / period
        out: list[float | None] = [None] * (period - 1) + [seed]
        for i in range(period, n):
            out.append(values[i] * k + out[i - 1] * (1 - k))
        return out  # type: ignore[return-value]
    out_simple = [values[0]]
    for i in range(1, n):
        out_simple.append(values[i] * k + out_simple[i - 1] * (1 - k))
    return out_simple


def rsi_series(values: list[float], period: int = 14) -> list[float | None]:
    """RSI via Wilder's smoothed moving average."""
    n = len(values)
    if n < period + 1:
        return [None] * n
    out: list[float | None] = [None] * n
    gains = []
    losses = []
    for i in range(1, period + 1):
        diff = values[i] - values[i - 1]
        gains.append(max(diff, 0.0))
        losses.append(max(-diff, 0.0))
    avg_gain = sum(gains) / period
    avg_loss = sum(losses) / period
    if avg_loss < 1e-12:
        out[period] = 100.0
    else:
        out[period] = 100.0 - 100.0 / (1.0 + avg_gain / avg_loss)
    for i in range(period + 1, n):
        diff = values[i] - values[i - 1]
        avg_gain = (avg_gain * (period - 1) + max(diff, 0.0)) / period
        avg_loss = (avg_loss * (period - 1) + max(-diff, 0.0)) / period
        if avg_loss < 1e-12:
            out[i] = 100.0
        else:
            out[i] = 100.0 - 100.0 / (1.0 + avg_gain / avg_loss)
    return out


def bollinger_at(values: list[float], period: int, mult: float) -> dict[str, Any] | None:
    """计算最后一根 bar 的 Bollinger Band 指标。"""
    n = len(values)
    if n < period or period < 2:
        return None
    window = values[n - period:]
    sma = sum(window) / period
    var = sum((x - sma) ** 2 for x in window) / period
    std = math.sqrt(var)
    upper = sma + mult * std
    lower = sma - mult * std
    price = values[-1]
    band_width = upper - lower
    pct_b = (price - lower) / band_width if band_width > 1e-12 else 0.5
    bandwidth = band_width / sma * 100 if sma > 0 else 0.0
    return {
        "upper": upper,
        "middle": sma,
        "lower": lower,
        "percentB": pct_b,
        "bandwidth": bandwidth,
    }


def bollinger_series(values: list[float], period: int, mult: float) -> list[dict | None]:
    """Rolling Bollinger Band for all bars, returns list aligned with values."""
    n = len(values)
    out: list[dict | None] = [None] * n
    if n < period or period < 2:
        return out
    for i in range(period - 1, n):
        window = values[i - period + 1: i + 1]
        sma = sum(window) / period
        var = sum((x - sma) ** 2 for x in window) / period
        std = math.sqrt(var)
        upper = sma + mult * std
        lower = sma - mult * std
        bw = upper - lower
        pct_b = (values[i] - lower) / bw if bw > 1e-12 else 0.5
        bandwidth = bw / sma * 100 if sma > 0 else 0.0
        out[i] = {"percentB": pct_b, "bandwidth": bandwidth}
    return out


def _fuse_with_bb(base_signal: str, pct_b: float, bw_expanding: bool,
                  buy_kill: float = 0.3, sell_kill: float = 0.7) -> str:
    """
    用 Bollinger %B 位置和带宽变化修正 EMA 动量信号。

    buy_kill / sell_kill 可按品种配置，降低阈值可减少压制、提升灵敏度。

    规则：
    - buy  + %B < buy_kill    → neutral     位置与方向矛盾
    - buy  + %B > 0.5 + 扩张  → strong_buy  趋势/位置/波动率三确认
    - strong_buy + %B > 1.0   → buy         过度延伸
    - sell + %B > sell_kill    → neutral     位置与方向矛盾
    - sell + %B < 0.5 + 扩张  → strong_sell 三确认
    - strong_sell + %B < 0.0  → sell        超卖反弹风险
    """
    sig = base_signal
    if sig == "buy":
        if pct_b < buy_kill:
            sig = "neutral"
        elif pct_b > 0.5 and bw_expanding:
            sig = "strong_buy"
    elif sig == "strong_buy":
        if pct_b > 1.0:
            sig = "buy"
    elif sig == "sell":
        if pct_b > sell_kill:
            sig = "neutral"
        elif pct_b < 0.5 and bw_expanding:
            sig = "strong_sell"
    elif sig == "strong_sell":
        if pct_b < 0.0:
            sig = "sell"
    return sig


def _fuse_with_rsi(base_signal: str, rsi: float) -> str:
    """RSI 超买/超卖修正信号。"""
    sig = base_signal
    if sig == "buy" and rsi > 70:
        sig = "neutral"
    elif sig == "sell" and rsi < 30:
        sig = "neutral"
    return sig


def _apply_signal_cooldown(signal: str, last_active: str, cooldown: int, cooldown_bars: int) -> tuple[str, int, bool]:
    """
    实时信号 cooldown 逻辑：方向翻转后 N 个 bar 内压制信号为 neutral。

    规则：
    - 趋势持续（buy→buy, buy→strong_buy）：信号正常输出，cooldown 自然衰减
    - 趋势反转（buy→sell）：cooldown 期间被压制为 neutral
    - 趋势减弱（buy→neutral）：正常输出 neutral
    - 新方向信号（neutral→buy/sell）触发 cooldown

    返回: (处理后的信号, 新的 cooldown 值, 是否更新了 last_active)
    """
    def _direction(s: str) -> str:
        if s in ("buy", "strong_buy"):
            return "long"
        if s in ("sell", "strong_sell"):
            return "short"
        return "flat"

    sig_dir = _direction(signal)
    last_dir = _direction(last_active)

    new_cooldown = max(0, cooldown - 1)
    updated_last = False

    # cooldown 期间，只允许同向信号通过；反向信号被压制为 neutral
    if cooldown > 0 and sig_dir != last_dir and sig_dir != "flat":
        return "neutral", new_cooldown, updated_last

    # 新的活跃方向信号触发 cooldown（包括 flat→long/short，以及 long↔short）
    if sig_dir != "flat" and sig_dir != last_dir:
        new_cooldown = cooldown_bars

    if sig_dir != "flat":
        updated_last = True

    return signal, new_cooldown, updated_last


def calc_momentum(vals: list[float], params: MomentumParams | None = None) -> dict[str, Any] | None:
    """
    输入收盘价序列（与 JS 中 series 的 y 一致）。
    返回与 momentum.js calcMomentum 相同语义的字段；样本不足时返回 None。
    当 bb_period > 0 时自动计算 Bollinger 带并融合信号。
    """
    p = params or MomentumParams()
    min_len = p.long_p + 2
    if not vals or len(vals) < min_len:
        return None

    ema_s = ema_series(vals, p.short_p)
    ema_l = ema_series(vals, p.long_p)
    last_s = ema_s[-1]
    last_l = ema_l[-1]
    prev_s = ema_s[-2]
    if last_s is None or last_l is None or prev_s is None:
        return None
    spread_pct = ((last_s - last_l) / last_l) * 100 if last_l != 0 else 0.0
    slope_pct = ((last_s - prev_s) / prev_s) * 100 if prev_s != 0 else 0.0

    # EMA 基础信号
    signal = "neutral"
    if last_s > last_l and spread_pct > p.spread_entry and slope_pct > p.slope_entry:
        signal = "strong_buy" if spread_pct > p.spread_strong else "buy"
    elif last_s < last_l and spread_pct < -p.spread_entry and slope_pct < -p.slope_entry:
        signal = "strong_sell" if spread_pct < -p.spread_strong else "sell"

    # Bollinger 带融合
    bb_info: dict[str, Any] | None = None
    squeeze = False
    if p.bb_period > 0 and len(vals) >= p.bb_period:
        bb_now = bollinger_at(vals, p.bb_period, p.bb_mult)
        bb_prev = bollinger_at(vals[:-1], p.bb_period, p.bb_mult) if len(vals) > p.bb_period else None
        if bb_now:
            bw_expanding = bb_prev is not None and bb_now["bandwidth"] > bb_prev["bandwidth"]
            signal = _fuse_with_bb(signal, bb_now["percentB"], bw_expanding,
                                   buy_kill=p.bb_buy_kill, sell_kill=p.bb_sell_kill)
            # Squeeze: 当前带宽 ≤ 近 bb_period 根 bar 最小带宽
            if len(vals) >= p.bb_period * 2:
                bws: list[float] = []
                for j in range(p.bb_period):
                    end = len(vals) - j
                    if end >= p.bb_period:
                        bb_j = bollinger_at(vals[:end], p.bb_period, p.bb_mult)
                        if bb_j:
                            bws.append(bb_j["bandwidth"])
                if bws and bb_now["bandwidth"] <= min(bws):
                    squeeze = True
            bb_info = {
                "upper": bb_now["upper"],
                "middle": bb_now["middle"],
                "lower": bb_now["lower"],
                "percentB": bb_now["percentB"],
                "bandwidth": bb_now["bandwidth"],
                "bwExpanding": bw_expanding,
                "squeeze": squeeze,
            }

    # RSI 融合
    rsi_val: float | None = None
    if p.rsi_period > 0 and len(vals) >= p.rsi_period + 1:
        rsi_all = rsi_series(vals, p.rsi_period)
        rsi_val = rsi_all[-1]
        if rsi_val is not None:
            signal = _fuse_with_rsi(signal, rsi_val)

    # 波动率过滤：横盘/噪声环境时抑制方向性信号
    # 自适应阈值：根据近期 CV 动态调整，低波动时逐步放开
    volatility_pct: float | None = None
    adaptive_vol_threshold: float | None = None
    if p.min_volatility_pct > 0.0 and len(vals) >= max(p.bb_period, p.short_p):
        lookback = max(p.bb_period, p.short_p)
        recent = vals[-lookback:]
        avg = sum(recent) / len(recent)
        if avg > 0:
            variance = sum((x - avg) ** 2 for x in recent) / len(recent)
            cv = math.sqrt(variance) / avg * 100.0
            volatility_pct = cv
            # 自适应阈值逻辑
            if cv < 0.005:
                adaptive_vol_threshold = 0.0  # 超超低波动：完全放开
            elif cv < 0.015:
                adaptive_vol_threshold = cv * 0.3  # 超低波动：极宽松
            elif cv < 0.03:
                adaptive_vol_threshold = cv * 0.6  # 低波动：宽松
            else:
                adaptive_vol_threshold = p.min_volatility_pct  # 正常波动：使用配置值
            if cv < adaptive_vol_threshold and signal in ("buy", "strong_buy", "sell", "strong_sell"):
                signal = "neutral"

    strength = min(100.0, abs(spread_pct) * p.strength_multiplier)

    result: dict[str, Any] = {
        "signal": signal,
        "spreadPct": spread_pct,
        "slopePct": slope_pct,
        "shortEMA": last_s,
        "longEMA": last_l,
        "strength": strength,
    }
    if bb_info:
        result["bb"] = bb_info
    if rsi_val is not None:
        result["rsi"] = rsi_val
    if volatility_pct is not None:
        result["volatilityPct"] = round(volatility_pct, 4)
    if adaptive_vol_threshold is not None:
        result["adaptiveVolThreshold"] = round(adaptive_vol_threshold, 4)
    return result
